### PR TITLE
Add component dialog

### LINKED ISSUE #110: Detailed editing of entire NeXus Structure
I don't see any way around having an editable treeview for the entire NeXus structure, but other suggestions are welcome.
A tab setup to switch between the basic component list and a full treeview seems like a nice idea.
This ticket needs to be discussed and planned in more detail before starting any coding!

### PR DESCRIPTION
**Pull request by rerpha**
### Issue

Closes #303 
Closes #110

### Description of work

Added an "add component" dialog. 

This should work and add components into the NeXus file as well as the component model. Transformations are not in this window. 

### Acceptance Criteria 

Add a few components and export to nexus file, then check the components have been added in the nexus file. There isn't really a way of checking the instrument model easily without going in with a debugger but to test you could put a breakpoint on somewhere after it has added the component and check it exists in the instrument model. 

### UI tests

N/A

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
Note: one needs to `git submodule update --init` for this branch to work.

File: nexus_constructor/addcomponentwindow.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
+from enum import Enum
```
Comment (matthew-d-jones) [inline nexus_constructor/addcomponentwindow.py]: Can we use snake_case for file names?

File: nexus_constructor/addcomponentwindow.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        self.component_types = make_dictionary_of_class_definitions(
+            os.path.abspath(os.path.join(os.curdir, "definitions"))
+        )
+
+    def setupUi(self, parent_dialog):
+        """ Sets up push buttons and validators for the add component window. """
+        super().setupUi(parent_dialog)
+
+        # Connect the button calls with functions
+        self.buttonBox.button(QDialogButtonBox.Ok).clicked.connect(self.on_ok)
```
Comment (matthew-d-jones) [inline nexus_constructor/addcomponentwindow.py]: Are these camelCase names created by NeXus Designer, or can we change them to snake_case?

File: nexus_constructor/component_type.py [hunk 3]
```diff
@@ -14,3 +17,37 @@ class ComponentType(Enum):
...
+
+
+def __list_base_class_files(repo_directory):
+    base_class_dir = os.path.join(repo_directory, "base_classes")
+    for file in os.listdir(base_class_dir):
+        if file.endswith(".nxdl.xml"):
+            yield os.path.join(base_class_dir, file)
+
+
+def make_dictionary_of_class_definitions(
```
Comment (matthew-d-jones) [inline nexus_constructor/component_type.py]: Could this be unit tested, or refactored to make unit testable?

**Comment by matthew-d-jones** [comment]
"Pixel options" should only be displayed for NXmonitor, NXdetector and NXdetector_module.
The default vector for cylinder should be x=0, y=0, z=1

File: nexus_constructor/addcomponentwindow.py (hunk 2 above)
Comment (rerpha) [reply nexus_constructor/addcomponentwindow.py]: They're created by Qt Designer, but I think we can change them to snake_case. I think Qt UI files are by default camelCase, not sure if there's a standard.

File: nexus_constructor/add_component_window.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+from nexus_constructor.validators import UnitValidator, NameValidator
+from nexus_constructor.nexus_wrapper import NexusWrapper
+from nexus_constructor.utils import file_dialog, validate_line_edit
+import os
+from functools import partial
+
+GEOMETRY_FILE_TYPES = {"OFF Files": ["off", "OFF"], "STL Files": ["stl", "STL"]}
+
+
+class FileValidator(QValidator):
```
Comment (matthew-d-jones) [inline nexus_constructor/add_component_window.py]: I suggest naming it more specifically `GeometryFileValidator` and moving it to its own file.

File: nexus_constructor/add_component_window.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        if self.meshRadioButton.isChecked():
+            geometry_model = OFFModel()
+            geometry_model.set_units(self.unitsLineEdit.text())
+            geometry_model.set_file(self.geometry_file_name)
+        else:
+            geometry_model = NoShapeModel()
+        return geometry_model
+
+    def on_ok(self):
+        component_type = self.componentTypeComboBox.currentText()
```
Comment (matthew-d-jones) [inline nexus_constructor/add_component_window.py]: Does "type" here mean "Nx_class"? If so let's use the terminology `nx_class` to reduce ambiguity.

File: nexus_constructor/component.py [hunk 6]
```diff
@@ -1,74 +1,19 @@
...
-        ),
-    )
 
 
 @attr.s
 class Component:
     """Components of an instrument"""
 
-    component_type = attr.ib(ComponentType)
+    nx_class = attr.ib(str)
```
Comment (matthew-d-jones) [inline nexus_constructor/component.py]: Good, I think there are other places that would benefit from this change too.

File: nexus_constructor/component_type.py [hunk 7]
```diff
@@ -1,16 +1,46 @@
...
+    class_definitions[nx_class_name] = []
+    try:
+        fields = xml_definition["field"]
+        try:
+            for field in fields:
+                class_definitions[nx_class_name].append(field["@name"])
+        except Exception:
+            class_definitions[nx_class_name].append(fields["@name"])
+    except KeyError:
+        # TODO: No key called "field"
```
Comment (matthew-d-jones) [inline nexus_constructor/component_type.py]: There is nothing "TODO" here, if there is no key called "field" then that is fine.

File: nexus_constructor/validators.py [hunk 8]
```diff
@@ -32,27 +32,25 @@ def validate(self, input: str, pos: int):
...
             return QValidator.Intermediate
 
-        self.validationSuccess.emit()
+        self.isValid.emit(True)
         return QValidator.Acceptable
 
-    # Create signals because the QML LabeledTextField doesn't have "onAccepted"
-    validationSuccess = Signal()
-    validationFailed = Signal()
+    isValid = Signal(bool)
```
Comment (matthew-d-jones) [inline nexus_constructor/validators.py]: Is this a standard signal name which must be camelCase, or can we use snake_case?

Comment (rerpha) [reply nexus_constructor/validators.py]: we can use snake_case here i believe.

**Pull request merged**